### DISCUSSION TITLE
How to use the SukiUI with ISingleViewApplicationLifetime

### DISCUSSION THREAD

**Discussion by sn4k3** [kikipoulet/SukiUI#235] [Q&A]
Hello,

I've an application that use `ISingleViewApplicationLifetime` and linux framebuffer to render the app under terminal (no desktop) linux. The application is based on UserControl as top level rather a Window.

As I see the SukiUI requires an SukiWindow for controls to render nicely. There is anyway to apply the styles of a Window to a UserControl?

I tried this hack:

```C#
public partial class MainView : UserControl
{
    protected override Type StyleKeyOverride => typeof(SukiWindow);
    public MainView()
    {
        InitializeComponent();
    }

    protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
    {
        base.OnApplyTemplate(e);
        var panel = e.NameScope.Find<LayoutTransformControl>("PART_LayoutTransform")?.Parent?.Parent as Panel;

        panel?.Children.Clear();
    }
}
```

But on linux it raises the error:

```
System.InvalidOperationException: SukiHost must be hosted inside a Window or SukiWindow
   at SukiUI.Controls.SukiHost.OnApplyTemplate(TemplateAppliedEventArgs e)
   at Avalonia.Controls.Primitives.TemplatedControl.ApplyTemplate()
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Layout.LayoutHelper.MeasureChild(Layoutable control, Size availableSize, Thickness padding)
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Layout.Layoutable.MeasureOverride(Size availableSize)
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Layout.LayoutHelper.MeasureChild(Layoutable control, Size availableSize, Thickness padding, Thickness borderThickness)
   at Avalonia.Controls.Border.MeasureOverride(Size availableSize)
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Layout.Layoutable.MeasureOverride(Size availableSize)
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Layout.LayoutHelper.MeasureChild(Layoutable control, Size availableSize, Thickness padding, Thickness borderThickness)
   at Avalonia.Controls.Presenters.ContentPresenter.MeasureOverride(Size availableSize)
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Layout.LayoutManager.Measure(Layoutable control)
   at Avalonia.Layout.LayoutManager.ExecuteMeasurePass()
   at Avalonia.Layout.LayoutManager.InnerLayoutPass()
   at Avalonia.Layout.LayoutManager.ExecuteLayoutPass()
   at Avalonia.Media.MediaContext.FireInvokeOnRenderCallbacks()
   at Avalonia.Media.MediaContext.RenderCore()
   at Avalonia.Media.MediaContext.Render()
   at Avalonia.Threading.DispatcherOperation.InvokeCore()
   at Avalonia.Threading.DispatcherOperation.Execute()
   at Avalonia.Threading.Dispatcher.ExecuteJob(DispatcherOperation job)
   at Avalonia.Threading.Dispatcher.ExecuteJobsCore(Boolean fromExplicitBackgroundProcessingCallback)
   at Avalonia.Threading.Dispatcher.Signaled()
   at Avalonia.Controls.Platform.ManagedDispatcherImpl.RunLoop(CancellationToken token)
   at Avalonia.Threading.DispatcherFrame.Run(IControlledDispatcherImpl impl)
   at Avalonia.Threading.Dispatcher.PushFrame(DispatcherFrame frame)
   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken)
   at LinuxFramebufferPlatformExtensions.StartLinuxDirect(AppBuilder builder, String[] args, IOutputBackend outputBackend, IInputBackend inputBackend)
   at Desktop.Program.Main(String[] args)
```

**Comment by kikipoulet** [comment]
I would say that using SukiUI for a mobile type application is not the best idea honestly. I have a library very similar to SukiUI for mobile btw : https://github.com/kikipoulet/CherylUI

If you really want to use the SukiUI system, you can check the SukiWindow code and recreate it at the root of your app without the window system.

https://github.com/kikipoulet/SukiUI/blob/main/SukiUI/Controls/SukiWindow.axaml


Basically you will just need to use the SukiBackground control if you want to use our background, and to be able to use Dialogs you will have to add the SukiHost control .

**Reply by sn4k3** [reply]
Thanks I will have a look into CheryIUI. But would be nice to be able to use Suki in both contexts, in my case my app execute as singlepage in linux framebuffer, but also as normal Window application under Windows and desktops.
If we can have a kind of `SukiUIRootUserControl` so we can use it as toplevel but also falldown to SukiWindows on desktops under same app.
